### PR TITLE
Adding a null HAL driver.

### PR DESCRIPTION
**Pull request by benvanik**
This is a dummy that can be used as a template and documentation for new HAL drivers and a change detector for HAL APIs. It's also useful as a way to get the minimum size of a HAL driver with default emulation.

### CODE REVIEW

File: runtime/src/iree/hal/drivers/null/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+DO NOT SUBMIT
+
+{Null} Vulkan
+_null_ _vulkan_
+TODO(null):
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/null/README.md]: 👀

File: runtime/src/iree/hal/drivers/null/allocator.c [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+#include "iree/hal/drivers/null/allocator.h"
+
+#include "iree/hal/drivers/null/buffer.h"
+
+// TODO(null): use one ID per address space or pool - each shows as a different
+// track in tracing tools.
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_ALLOCATION_TRACKING
+static const char* IREE_HAL_NULL_ALLOCATOR_ID = "{Null} unpooled";
+#endif  // IREE_TRACING_FEATURE_ALLOCATION_TRACKING
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/null/allocator.c:15]: 😂 `TODO(null)`

Are these TODOs something you want resolved here, or are you using them as a template for implementors?

File: CMakeLists.txt [hunk 3]
```diff
@@ -275,8 +275,9 @@ option(IREE_HAL_DRIVER_CUDA "Enables the 'cuda' runtime HAL driver" ${IREE_HAL_D
 option(IREE_HAL_DRIVER_HIP "Enables the 'hip' runtime HAL driver" ${IREE_HAL_DRIVER_HIP_DEFAULT})
 option(IREE_HAL_DRIVER_LOCAL_SYNC "Enables the 'local-sync' runtime HAL driver" ${IREE_HAL_DRIVER_DEFAULTS})
 option(IREE_HAL_DRIVER_LOCAL_TASK "Enables the 'local-task' runtime HAL driver" ${IREE_HAL_DRIVER_DEFAULTS})
-option(IREE_HAL_DRIVER_VULKAN "Enables the 'vulkan' runtime HAL driver" ${IREE_HAL_DRIVER_VULKAN_DEFAULT})
 option(IREE_HAL_DRIVER_METAL "Enables the 'metal' runtime HAL driver" ${IREE_HAL_DRIVER_METAL_DEFAULT})
+option(IREE_HAL_DRIVER_NULL "Enables the 'null' runtime HAL driver" ${IREE_HAL_DRIVER_DEFAULTS})
```
Comment (ScottTodd) [inline CMakeLists.txt]: Should we disable this in our release builds (or maybe by default when the CMake build type is `Release`), to save a bit of binary size? Hopefully applications can also rely on LTO to eliminate it.

**Review by ScottTodd — COMMENTED** [review]
Nice!

File: CMakeLists.txt (hunk 3 above)
Comment (benvanik) [reply CMakeLists.txt]: good call

File: runtime/src/iree/hal/drivers/null/allocator.c (hunk 2 above)
Comment (benvanik) [reply runtime/src/iree/hal/drivers/null/allocator.c:15]: they are templates - will push the readme in a sec

File: CMakeLists.txt (hunk 3 above)
Comment (benvanik) [reply CMakeLists.txt]: (so long as we have at least one bot building it - we have debug bots, right?)

Comment (benvanik) [reply CMakeLists.txt]: oo IREE_DEV_MODE works too

**Comment by benvanik** [comment]
bazel deps tracking saves the day again

File: CMakeLists.txt (hunk 3 above)
Comment (ScottTodd) [reply CMakeLists.txt]: We do have a debug build that runs nightly, yes. May be able to get that running on every commit if we configure a cache server (https://github.com/iree-org/iree/issues/18557) and have budget for the increased load.

Looks like `IREE_DEV_MODE` isn't tested via automation but it is a good fit for this.

File: runtime/src/iree/hal/drivers/null/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+This is a skeleton HAL driver that implements most interfaces needed by a HAL
+driver. It contains notes that can be useful to implementers, stubs for
+easy copy/paste/renaming, and some default behavior in places that many
+implementations can use (such as file queue operations). It doesn't run anything
+and nearly all methods return `UNIMPLEMENTED` errors: this can be used to
+incrementally build a driver by running until you hit an UNIMPLEMENTED,
+implementing it, and running again. Note however that the HAL is fairly regular
+and starting with the files and running down the methods is often much easier
+to do than the trial-and-error approach: if you can implement command buffer
+fill (memset) you can often implement copy (memcpy) as well at the same time.
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/null/README.md:12]: Could also add a section on testing.

* Example commands for compiling and running, for "running until you hit an UNIMPLEMENTED"
* Template CTS CMake file, like https://github.com/iree-org/iree/blob/main/runtime/src/iree/hal/drivers/metal/cts/CMakeLists.txt with all tests under `EXCLUDED_TESTS` to start

**Review by ScottTodd — APPROVED** [review]

File: runtime/src/iree/hal/drivers/null/README.md (hunk 4 above)
Comment (benvanik) [reply runtime/src/iree/hal/drivers/null/README.md:12]: in the future - I think we need to rework the CTS a bit - for now I'm ok with people needing to know how to use IREE to run anything in order to implement a HAL backend :)

**Comment by makslevental** [comment]
I'm Johnny-come-lately over here but is there any fear/concern here of a [bobby-tables](https://xkcd.com/327/) type situation - i.e., that `null` gets parsed somewhere/somehow?

**Comment by benvanik** [comment]
what is this, javascript? ;P
haven't run into any issues in the past - there's no issues in the language as we always namespace everything (`iree_hal_null_...`) and users best not be doing crazy automatic string conversions on their driver names in config dictionaries - that evil is on them if so :P

**Comment by benvanik** [comment]
(it's not compiled in release builds and doesn't do anything, so it'll be rare to be an issue for anyone but us devs)

**Pull request merged**